docs/src/prototypes/MenuButton/MenuButton.tsx

```diff
@@ -0,0 +1,198 @@
+  button: ShorthandValue<ButtonProps>
+  buttonId: string
+  disabled?: boolean
+  menu: ShorthandValue<MenuProps>
+  menuId: string
+  placement?: PopperJS.Placement
+}
+
+export interface MenuButtonState {
+  lastKeyCode: null | number
+  lastShiftKey: boolean
+  menuOpen: boolean
+}
+
+class MenuButton extends React.Component<MenuButtonProps, MenuButtonState> {
+  static defaultProps = {
+    accessibility: menuButtonBehavior,
+    placement: 'bottom',
+  }
+
+  state: MenuButtonState = {
+    lastKeyCode: null,
+    lastShiftKey: false,
+    menuOpen: false,
+  }
+
+  buttonRef = React.createRef<HTMLButtonElement>()
+  menuRef = React.createRef<HTMLUListElement>()
+
+  componentDidUpdate(_, prevState: MenuButtonState) {
+    if (!prevState.menuOpen && this.state.menuOpen) {
+      document.addEventListener('click', this.handleDocumentClick)
+
+      focusMenuItem(this.menuRef, this.state.lastKeyCode === keyboardKey.ArrowUp ? 'last' : 'first')
+    }
+
+    if (prevState.menuOpen && !this.state.menuOpen) {
+      document.removeEventListener('click', this.handleDocumentClick)
+
+      switch (this.state.lastKeyCode) {
+        case keyboardKey.Enter:
+        case keyboardKey.Escape:
+          focusButton(this.buttonRef)
+          break
+
+        case keyboardKey.Tab:
+          focusNearest(this.buttonRef, this.state.lastShiftKey ? 'previous' : 'next')
+          break
+      }
+    }
+  }
+
+  handleButtonOverrides = (predefinedProps: ButtonProps) => ({
+    onClick: (e: React.SyntheticEvent, buttonProps: ButtonProps) => {
+      _.invoke(predefinedProps, 'onClick', e, buttonProps)
+      this.setState(prevState => ({
+        lastKeyCode: null,
+        menuOpen: !prevState.menuOpen,
+      }))
+    },
+  })
+
+  handleDocumentClick = (e: MouseEvent) => {
+    const { menuOpen } = this.state
+    const target = e.target as HTMLElement
+    const isInside =
+      _.invoke(this.buttonRef.current, 'contains', target) &&
+      _.invoke(this.menuRef.current, 'contains', target)
+
+    if (menuOpen && !isInside) {
+      this.setState({ lastKeyCode: null, menuOpen: false })
+    }
+  }
+
+  handleKeyDown = (e: React.KeyboardEvent) => {
+    const { menuOpen } = this.state
+
+    const keyCode = keyboardKey.getCode(e)
+    const shouldClose =
+      menuOpen && _.includes([keyboardKey.Enter, keyboardKey.Escape, keyboardKey.Tab], keyCode)
+    const shouldOpen =
+      !menuOpen && _.includes([keyboardKey.ArrowDown, keyboardKey.ArrowUp], keyCode)
+
+    if (shouldClose) {
+      // We should prevent default there otherwise event will be bubbled to a button and will cause menu reopen
+      if (keyCode === keyboardKey.Enter) e.preventDefault()
+
+      this.setState({
+        lastKeyCode: keyCode,
+        lastShiftKey: e.shiftKey,
+        menuOpen: false,
+      })
+    }
+
+    if (shouldOpen) {
+      this.setState({ lastKeyCode: keyCode, menuOpen: true })
+    }
+  }
+
+  handleMenuItemOverrides = accessibilityBehavior =>
```

Comment:
lets accept `menuItemAccessibilityBehavior` as an argument of this method: this will decouple it from the parent accessibility object of `menuButton`, as well as will make it easier to infer the intent at the place of usage:

```js
Menu.create(menu, {
   defaultProps: {
      ...accessibilityBehavior.attributes.menu,
     overrideProps: {
       items: this.handleMenuItemOverrides(accessibilityBehavior.menuItem),
       ...
```

Reply:
Reworked this part, now we're passing: `(menuItemAccessibilityAttributes: AccessibilityAttributes) => {}`